lang/src/OTS/ITF/org/partiql/ots/operator/ScalarOp.kt

```diff
@@ -1,12 +1,40 @@
 package OTS.ITF.org.partiql.ots.operator
 
 import OTS.ITF.org.partiql.ots.CompileTimeType
+import OTS.ITF.org.partiql.ots.TypeInferenceResult
+import OTS.ITF.org.partiql.ots.type.BoolType
+import OTS.ITF.org.partiql.ots.type.ScalarType
 
 interface ScalarOp {
+    /**
+     * Used to check data type mismatch error for any operand.
+     */
+    // TODO: Will be removed after we support function overloading
+    val validOperandTypes: List<ScalarType>
+
     /**
      * Return type when we cannot decide the return type (type inference result is [Failed])
      *
      * Empty list leads to type inference result as MISSING
      */
     val defaultReturnTypes: List<CompileTimeType>
+
+    /**
+     * Infer the return type
+     */
+    fun inferReturnType(argsType: List<CompileTimeType>): TypeInferenceResult
+}
+
+abstract class PosOp : ScalarOp
+abstract class NegOp : ScalarOp
+abstract class NotOp : ScalarOp
+abstract class BinaryPlusOp : ScalarOp
+abstract class BinaryMinusOp : ScalarOp
+abstract class BinaryTimesOp : ScalarOp
+abstract class BinaryDivideOp : ScalarOp
+abstract class BinaryModuloOp : ScalarOp
+abstract class BinaryConcatOp : ScalarOp
+abstract class LikeOp : ScalarOp {
+    override val defaultReturnTypes: List<CompileTimeType>
+        get() = listOf(BoolType.compileTimeType)
 }
```

Comment:
I might overlook something, but these abstract classes inheriting from `ScalarOp` appear useless.  I believe, all their uses can be replaced with `ScalarOp` itself.  

[For another PR]
Now, if the above succeeds for the operators, it would be worthwhile to explore whether the types in the plugin can be architected similarly: 

- Note that singleton objects in `...ots.legacy.types` already inherit from a single interface `ScalarType`, without abstract types in between.  So, they are "ahead" of the operators in this regard. 

- However, they are "behind", in that `Plugin` and `StandardPlugin` do not list them as value fields, as they do for the operators.  I hope this can be done, which should eliminate the inappropriate references to them from core PartiQL (see, say, the import of Int2Type in StaticTypeInferenceVisitorTransform).

Reply:
Do you think we can add `operatorId` to `ScalarOp` and remove these abstract classes?